### PR TITLE
Slight platforms refactor

### PR DESCRIPTION
**Pull request by MetRonnie**
This is a small change with no associated Issue.

Move validation-y parts of `get_platform()` to their own functions. Changes should help improve type safety.

**Requirements check-list**
- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Appropriate tests are included (unit and/or functional).
- [x] No change log entry required (invisible to users).
- [x] No documentation update required.
- [x] No dependency changes.

### CODE REVIEW

File: cylc/flow/cfgspec/suite.py [hunk 1]
```diff
@@ -1604,12 +1606,27 @@ def upgrade_param_env_templates(cfg, descr):
...
 def warn_about_depr_platform(cfg):
-    """Warn if deprecated host or batch system appear in config."""
-    if 'runtime' in cfg:
-        for task_name, task_cfg in cfg['runtime'].items():
-            platform = get_platform(task_cfg, task_name, warn_only=True)
-            if type(platform) == str:
-                LOG.warning(platform)
+    """Validate platforms config.
+
+    - Warn if deprecated host or batch system appear in config,
```
Comment (kinow) [inline cylc/flow/cfgspec/suite.py:1611]: Nit-pick, `,` only here, `.` elsewhere :rofl:

File: tests/unit/test_platforms_get_platform.py [hunk 2]
```diff
@@ -196,23 +196,6 @@ def test_get_platform_using_platform_from_job_info(
     assert get_platform(task_conf)['name'] == expected_platform_name
```
Comment (kinow) [inline tests/unit/test_platforms_get_platform.py:198]: @MetRonnie with this test I think the Codecov warning above would go away.

```py
def test_get_platform_from_config_with_subshell(mock_glbl_cfg):
    # A platform definition is using subshell:
    mock_glbl_cfg(
        'cylc.flow.platforms.glbl_cfg',
        '''
        [platforms]
            [[mace]]
                hosts = mace001, mace002
                job runner = slurm
        '''
    )
    task_conf = {'platform': '$(host)'}
    platform = get_platform(task_conf)
    assert platform is None
```

(Not sure if `mock_glbl_cfg` is necessary or needs changing too, but just to test that option that returns None) :+1:

**Review by kinow — APPROVED** [review]
Couple comments but no blockers :+1:

File: cylc/flow/cfgspec/suite.py [hunk 3]
```diff
@@ -1604,12 +1606,27 @@ def upgrade_param_env_templates(cfg, descr):
...
+    - Raise if platforms section is also present.
```
Comment (wxtim) [inline cylc/flow/cfgspec/suite.py:1612]: You know that the reason that this wasn't here is that not all instances where this conflict happens will appear here? Some conflicts don't appear until after task inheritance is expanded.

File: cylc/flow/platforms.py [hunk 4]
```diff
@@ -48,105 +46,66 @@
...
+            this should be a configuration for a task.
+        task_id: Task identification string - help produce more helpful error
             messages.
-        warn_only(bool):
-            If true, warnings about tasks requiring upgrade will be returned.
 
     Returns:
-        platform (platform, or string):
-            Actually it returns either get_platform() or
+        platform: Actually it returns either get_platform() or
```
Comment (wxtim) [inline cylc/flow/platforms.py:65]: I don't like my original text - I think:

> platform: A platform definition object.

Because I think that the fuller explanation of the mechanics of platform selection are not required _here_.

File: cylc/flow/platforms.py [hunk 5]
```diff
@@ -48,105 +46,66 @@
...
-
     """
-    if task_conf is None:
-        # Just a simple way of accessing localhost items.
-        output = platform_from_name()
-
-    elif isinstance(task_conf, str):
-        # If task_conf is str assume that it is a platform name.
-        output = platform_from_name(task_conf)
+    if task_conf is None or isinstance(task_conf, str):
```
Comment (wxtim) [inline cylc/flow/platforms.py:68]: Very tidy.

**Review by wxtim — APPROVED** [review]
Good work.
- [x] Test passing
- [ ] Code Read
- [ ] Checked on some toy suites - particularly trying to break the code using inheritance.

File: tests/unit/test_platforms_get_platform.py (hunk 2 above)
Comment (MetRonnie) [reply tests/unit/test_platforms_get_platform.py:198]: Added here https://github.com/cylc/cylc-flow/pull/4152/commits/969f758dc176efc4e4d362a0b690c9ea62839dfa
